src/transformers/utils/type_validators.py

```diff
@@ -0,0 +1,213 @@
+from collections.abc import Sequence
+from dataclasses import MISSING, field, make_dataclass
+from typing import Annotated, ForwardRef, Optional, TypedDict, Union, get_args, get_origin
+
+from huggingface_hub.dataclasses import as_validated_field, strict
+
+from ..tokenization_utils_base import PaddingStrategy, TruncationStrategy
+from ..video_utils import VideoMetadataType
+from .generic import TensorType
+from .import_utils import is_vision_available
+
+
+if is_vision_available():
+    from ..image_utils import PILImageResampling
+
+
+def unpack_annotated_type(type):
+    if get_origin(type) is Annotated:
+        base, *meta = get_args(type)
+        return base, meta[0]
+    return type, field(default=MISSING)
+
+
+def get_type_hints_from_typed_dict(obj: type[TypedDict]):
+    """
+    Same as `typing.get_type_hints` but does not perform evaluation
+    on the ForwardRefs. Evaluating might fails if the package is not imported
+    or installed, therefore we will have our own "guarded" type validations.
+    All `ForwardRef` will be ignored by the hub validator
+    """
+    raw_annots = obj.__dict__.get("__annotations__", {})
+    type_hints = {}
+    for name, value in raw_annots.items():
+        if value is None:
+            value = type(None)
+        if isinstance(value, str):
+            value = ForwardRef(value, is_argument=False)
+        type_hints[name] = value
+    return type_hints
+
+
+# Minimalistic version of pydantic.TypeAdapter tailored for `TypedDict`
+class TypedDictAdapter:
```

Comment:
Sorry about late comment but I'm actively working in adding this back to `huggingface_hub` directly. Would it be ok to wait a bit before merging this PR? (if too annoying because of upcoming merge conflicts, then it's fine to merge and we will remove it from transformers later)